### PR TITLE
util/net utils: Implement file download resume.

### PR DESCRIPTION
**Pull request by Zangetsu38**
# About
util/net utils: Implement file download resume.
gui/vita3k update: Restore resume download new build.

### CODE REVIEW

File: vita3k/util/src/net_utils.cpp [hunk 1]
```diff
@@ -393,47 +393,58 @@ bool download_file(const std::string &url, const std::string &output_file_path,
...
 
         ProgressState *callback_result = data->second(progress_percent, remaining_time);
 
         std::unique_lock<std::mutex> lock(callback_result->mutex);
 
-        callback_result->cv.wait(lock, [&]() {
-            return !callback_result->pause;
-        });
+        if (callback_result->pause) {
+            callback_result->cv.wait(lock);
```
Comment (Macdu) [inline vita3k/util/src/net_utils.cpp:423]: Don't wait without a predicate, c++ condition variables are allowed to have a spurious wakeup

File: vita3k/util/src/net_utils.cpp [hunk 2]
```diff
@@ -393,47 +393,58 @@ bool download_file(const std::string &url, const std::string &output_file_path,
...
+
+            // Add the time spent paused to the total elapsed time
```
Comment (Macdu) [inline vita3k/util/src/net_utils.cpp:425]: Isn't it the opposing: "Remove to time spent paused..." instead (the code is right, just the comment means the opposite)

Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: no commentaire is fine, i no remove, i adding time elipsed in pause

File: vita3k/util/src/net_utils.cpp (hunk 1 above)
Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: yes, i have fixed it withlittle reworks this part

**Comment by Zangetsu38** [comment]
Pr is ready

File: vita3k/util/src/net_utils.cpp (hunk 2 above)
Comment (Macdu) [reply vita3k/util/src/net_utils.cpp]: You are increasing `data->first.time` which is effectively the starting time. The elapsed time is `now - data->first.time` which as a consequence decreases.
 What the line under does is "remove the time spent paused from the total elapsed time" which is indeed what you want to do for the remaining time approximation to be accurate.

Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: Current comment is little diferent but

The goal here is to keep track of the actual elapsed time excluding any periods where the download was paused. The data->first.time variable represents the moment the download started. When we pause the download, the time essentially 'stops,' so when we resume, we need to account for that gap in time to ensure the total elapsed time reflects only the periods when the download was actively in progress. By adding the difference between the current time and the time right before we paused (current_time), we're compensating for the pause duration. This way, the elapsed time will continue to reflect the correct duration from start to resume, excluding any pauses.

Comment (Macdu) [reply vita3k/util/src/net_utils.cpp]: I mean, the fact that you are giving a chatgpt-generated answer (or at least something clearly not by you) shows you don't understand why this comment is or is not misleading. I can do the same thing:
image

Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: I just used chatgpt to better explain to you what I wanted to say in English, he just helped me explain it correctly, he is also the one who confirmed to me that the comment posted perfectly explained what the code was doing, (the default comment was automatically written by copilot), and that you would have misunderstood the code.

I find it funny that they tell you that from your side when he literally explains the opposite to me.

But so to use my words, what the code does is when exiting the pause, it adds the time that has elapsed to the start time, so the comments explain this perfectly, the value of time increases, it does not decrease precisely

Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: For mor explain why i no can't using your version
Screenshot_20240919_223502_ChatGPT

Comment (Macdu) [reply vita3k/util/src/net_utils.cpp]: Are you really saying that given that the comment was written by copilot and chatgpt said it was correct, it is correct? ....

Also the AI generated response in your last message makes absolutely no sense.

Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: I'm not sure I understand what the problem is with the comment 🤔 
It's literally written what the code does, it adds to the starting time the time that elapsed during the break. 
Precisely to not take the break into account and to calculate the remaining time.

For copilot, he just made a suggestion once I had written the beginning "// Add" but then I rewrote this comment myself to be clearer.

Comment (Macdu) [reply vita3k/util/src/net_utils.cpp]: Yes, this exactly what you said:
- It adds to the starting time the time "elapsed during the break", i.e the time spent paused

Then why is your comment not mentioning the starting time and instead saying that the time spent paused is added to the total elapsed time? The total elapsed time can be computed as "now() - starting_time", so this comment is describing the opposite of what the code below is doing.

Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: Oh yeah, sorry...

Comment (antoniodesousa) [reply vita3k/util/src/net_utils.cpp]: I think Macdu wants the comment to be changed to `Add the time spent paused to the starting time`, which makes sense.

Comment (Zangetsu38) [reply vita3k/util/src/net_utils.cpp]: > I think Macdu wants the comment to be changed to `Add the time spent paused to the starting time`, which makes sense.

yes no worry, i have undersstund, and here i have no idea why i have confonded start time to elapsed time...

and i have juste removed elapsed time copilot have added alone start time... anyway, is fixed

**Review by Macdu — APPROVED** [review]

**Pull request merged**